lib/workers/repository/update/branch/execute-upgrade-commands.ts

```diff
@@ -0,0 +1,204 @@
+import is from '@sindresorhus/is';
+import minimatch from 'minimatch';
+import { mergeChildConfig } from '../../../../config';
+import { GlobalConfig } from '../../../../config/global';
+import type { PostUpgradeTasks } from '../../../../config/types';
+import { addMeta, logger } from '../../../../logger';
+import type { ArtifactError } from '../../../../modules/manager/types';
+import { exec } from '../../../../util/exec';
+import {
+  localPathIsFile,
+  readLocalFile,
+  writeLocalFile,
+} from '../../../../util/fs';
+import { getRepoStatus } from '../../../../util/git';
+import type { FileChange } from '../../../../util/git/types';
+import { regEx } from '../../../../util/regex';
+import { sanitize } from '../../../../util/sanitize';
+import { compile } from '../../../../util/template';
+import type { BranchConfig, BranchUpgradeConfig } from '../../../types';
+
+export interface UpgradeCommandsExecutionResult {
+  updatedArtifacts: FileChange[];
+  artifactErrors: ArtifactError[];
+}
+
+export async function upgradeTaskExecutor(
+  upgradeTask: PostUpgradeTasks | null | undefined,
+  config: BranchConfig,
+  updatedArtifacts: FileChange[],
+  allowedUpgradeCommands: string[] | undefined,
+  allowUpgradeCommandTemplating: boolean | undefined,
+  upgrade: BranchUpgradeConfig,
+  taskType: string
+): Promise<UpgradeCommandsExecutionResult> {
+  let currentUpdatedArtifacts = updatedArtifacts;
+  let artifactErrors: ArtifactError[] = [];
+
+  addMeta({ dep: upgrade.depName });
+  logger.trace(
+    {
+      tasks: upgradeTask,
+      allowedCommands: allowedUpgradeCommands,
+    },
+    `Checking for ${taskType.toLowerCase()} tasks`
+  );
+
+  const commands = [];
+  for (const cmd of upgradeTask?.commands ?? []) {
+    if (
+      (allowedUpgradeCommands ?? []).some((pattern) => regEx(pattern).test(cmd))
+    ) {
+      commands.push(cmd);
+    } else {
+      logger.warn(
+        {
+          cmd,
+          allowedUpgradeCommands,
+        },
+        `${taskType} task did not match any on allowedPostUpgradeCommands list`
+      );
+      artifactErrors.push({
+        lockFile: upgrade.packageFile,
+        stderr: sanitize(
+          `${taskType} command '${cmd}' has not been added to the allowed list in allowedPostUpgradeCommands`
+        ),
+      });
+    }
+  }
+
+  if (is.nonEmptyArray(commands)) {
+    // Persist updated files in file system so any executed commands can see them
+    const filesToPersist = (config.updatedPackageFiles ?? []).concat(
+      currentUpdatedArtifacts
+    );
+    await persistUpdatedFiles(filesToPersist);
+
+    for (const cmd of commands) {
+      const commandError = await upgradeCommandExecutor(
+        allowedUpgradeCommands ?? [],
+        cmd,
+        allowUpgradeCommandTemplating,
+        config,
+        upgrade,
+        taskType
+      );
+      artifactErrors = [...artifactErrors, ...commandError];
```

Comment:
use push instead, so we don't clone the array on each iteration .

i think we should better pass the existing array and push in called function to have less clones